pages/shared/studentGroupAssessmentHelpers/studentGroupAssessmentHelpers.js

```diff
@@ -0,0 +1,137 @@
+      }
+      next();
+    });
+  });
+};
+
+module.exports.joinGroup = function (req, res, next) {
+  try {
+    const group_name = req.body.join_code.split('-')[0];
+    const join_code = req.body.join_code.split('-')[1].toUpperCase();
+    if (join_code.length !== 4) {
+      throw 'invalid length of join code';
+    }
+    let params = [
+      res.locals.assessment.id,
+      res.locals.user.user_id,
+      res.locals.authn_user.user_id,
+      group_name,
+      join_code,
+    ];
+    sqldb.call('group_users_insert', params, function (err, _result) {
+      if (err) {
+        let params = {
+          assessment_id: res.locals.assessment.id,
+        };
+        sqldb.query(sql.get_config_info, params, function (err, result) {
+          if (ERR(err, next)) return;
+          res.locals.permissions = result.rows[0];
+          res.locals.groupsize = 0;
+          //display the error on frontend
+          res.locals.used_join_code = req.body.join_code;
+          //res.render(__filename.replace(/\.js$/, '.ejs'), res.locals);
+          next();
+          return;
+        });
+      } else {
+        res.redirect(req.originalUrl);
+      }
+    });
+  } catch (err) {
+    // the join code input by user is not valid (not in format of groupname+4-character)
+    let params = {
+      assessment_id: res.locals.assessment.id,
+    };
+    sqldb.query(sql.get_config_info, params, function (err, result) {
+      if (ERR(err, next)) return;
+      res.locals.permissions = result.rows[0];
+      res.locals.groupsize = 0;
+      //display the error on frontend
+      res.locals.used_join_code = req.body.join_code;
+      next();
+    });
+  }
+};
+
+module.exports.createGroup = function (req, res, next) {
+  const params = {
+    assessment_id: res.locals.assessment.id,
+    user_id: res.locals.user.user_id,
+    authn_user_id: res.locals.authn_user.user_id,
+    group_name: req.body.groupName,
+  };
+  //alpha and numeric characters only
+  let invalidGroupName = true;
+  const letters = /^[0-9a-zA-Z]+$/;
+  if (req.body.groupName.length <= 30 && req.body.groupName.match(letters)) {
+    invalidGroupName = false;
+    //try to create a group
+    sqldb.query(sql.create_group, params, function (err, _result) {
+      if (!err) {
+        res.redirect(req.originalUrl);
+      } else {
+        sqldb.query(sql.get_config_info, params, function (err, result) {
+          if (ERR(err, next)) return;
+          res.locals.permissions = result.rows[0];
+          res.locals.groupsize = 0;
+          res.locals.uniqueGroupName = true;
+          next();
+        });
+      }
+    });
+  }
+  if (invalidGroupName) {
+    sqldb.query(sql.get_config_info, params, function (err, result) {
+      if (ERR(err, next)) return;
+      res.locals.permissions = result.rows[0];
+      res.locals.groupsize = 0;
+      res.locals.invalidGroupName = true;
+      next();
+    });
+  }
+};
+module.exports.leaveGroup = function (req, res, next) {
```

Comment:
Ditto for `next` -> `callback` and removing the unused `req` parameter.

Reply:
Fixed partially in [4baf174](https://github.com/PrairieLearn/PrairieLearn/pull/5290/commits/4baf1740683c146a1b89ad57c707d5f3942bc555) and req removed in [bdb0919](https://github.com/PrairieLearn/PrairieLearn/pull/5290/commits/bdb0919247e79737a2ba6a75b11cbd7204381564)